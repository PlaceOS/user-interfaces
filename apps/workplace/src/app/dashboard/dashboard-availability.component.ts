import { Component, OnDestroy, OnInit } from '@angular/core';

import { DashboardStateService } from './dashboard-state.service';

@Component({
    selector: 'a-dashboard-availability',
    template: `
        <h3 class="m-0 font-medium text-xl">Available Now</h3>
        <div class="flex flex-col items-center p-2 bg-white my-4 rounded-lg">
            <div class="flex items-center w-full mb-2">
                <app-icon
                    class="bg-darken rounded-lg mr-2 text-2xl"
                    [icon]="{ class: 'material-icons', content: 'crop_free' }"
                ></app-icon>
                <div>Most free space:</div>
            </div>
            <div free-levels class="items space-y-2 w-full">
                <a
                    button
                    mat-button
                    class="w-full"
                    *ngFor="let lvl of levels_free | async | slice: 0:2"
                    [routerLink]="['/explore']"
                    [queryParams]="{ level: lvl.id }"
                >
                    {{ lvl.display_name || lvl.name }}
                </a>
            </div>
        </div>
        <div
            free-spaces
            class="flex items-center flex-col rounded-lg bg-white my-4 p-2"
        >
            <div class="flex items-center w-full mb-2">
                <app-icon
                    class="bg-darken rounded-lg mr-2 text-2xl"
                    [icon]="{ class: 'material-icons', content: 'av_timer' }"
                ></app-icon>
                <div>Free rooms now:</div>
            </div>
            <div class="items space-y-2">
                <a
                    button
                    mat-button
                    class="w-full"
                    *ngFor="let space of space_list | async | slice: 0:3"
                    [routerLink]="['/explore']"
                    [queryParams]="{ space: space.id }"
                    >{{ space.display_name || space.name }}
                </a>
                <span
                    *ngIf="!(space_list | async).length"
                    class="text-dark-fade text-sm mb-2"
                    >No free spaces</span
                >
            </div>
        </div>
    `,
    styles: [
        `
            :host {
                padding: 1em;
                min-width: 16rem;
            }
        `,
    ],
})
export class DashboardAvailabilityComponent implements OnInit, OnDestroy {
    public readonly space_list = this._state.free_spaces;
    public readonly levels_free = this._state.level_occupancy;

    constructor(private _state: DashboardStateService) {}

    public async ngOnInit() {
        this._state.pollFreeSpaces();
    }

    public ngOnDestroy() {
        this._state.stopPollingFreeSpaces();
    }
}
