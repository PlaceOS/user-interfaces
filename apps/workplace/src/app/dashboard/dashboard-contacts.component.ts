import { Component, ElementRef, ViewChild } from '@angular/core';
import { BaseClass, SettingsService } from '@placeos/common';

import { DashboardStateService } from './dashboard-state.service';

@Component({
    selector: 'a-dashboard-contacts',
    template: `
        <h3 class="m-0 mb-2 font-medium text-xl">Contacts</h3>
        <div
            name="contact-list"
            class="bg-gray-400 rounded-lg p-2 pl-0 space-y-2 flex flex-wrap"
        >
            <div
                name="contact"
                *ngFor="let user of contacts | async"
                class="bg-white rounded-lg flex-1 flex flex-col items-center p-2 relative"
            >
                <a-user-avatar
                    class="text-2xl mb-2"
                    [user]="user"
                ></a-user-avatar>
                <div name="contact-name" class="text-sm mb-2">
                    {{ user.name }}
                </div>
                <div name="location" class="flex items-center text-sm">
                    <div name="dot" class="mr-2"></div>
                    Level 12
                </div>
                <button mat-icon-button name="close" (click)="removeUser(user)">
                    <app-icon
                        [icon]="{ class: 'material-icons', content: 'close' }"
                    ></app-icon>
                </button>
            </div>
            <button
                name="contact"
                mat-button
                class="rounded-lg flex-1 w-full"
                [matMenuTriggerFor]="menu"
                (menuOpened)="focusInput()"
                (menuClosed)="clearInput()"
            >
                <div class="flex flex-col items-center p-4 text-black w-full">
                    <app-icon
                        class="text-2xl"
                        [icon]="{
                            class: 'material-icons',
                            content: 'person_add'
                        }"
                    ></app-icon>
                    <span class="w-full leading-tight whitespace-normal"
                        >Click to add a person to contacts</span
                    >
                </div>
            </button>
        </div>
        <mat-menu #menu="matMenu" xPosition="before">
            <div
                mat-menu-item
                class="p-0"
                (click)="$event.stopPropagation(); $event.preventDefault()"
            >
                <input
                    #input
                    class="w-full p-2 text-base"
                    [(ngModel)]="search_str"
                    (ngModelChange)="updateSearch(search_str)"
                    placeholder="Search for user..."
                />
            </div>
            <button mat-menu-item *ngIf="!(search_results | async)?.length">
                {{
                    !search_str
                        ? 'Type to search for users...'
                        : 'No matches for "' + search_str + '"'
                }}
            </button>
            <button
                mat-menu-item
                *ngFor="let user of search_results | async | slice: 0:8"
                (click)="addUser(user)"
            >
                <div class="flex items-center leading-tight">
                    <a-user-avatar
                        class="text-xl mr-2"
                        [user]="user"
                    ></a-user-avatar>
                    <div class="flex-1">
                        <div>{{ user.name }}</div>
                        <div class="text-xs text-dark-fade">
                            {{ user.email }}
                        </div>
                    </div>
                </div>
            </button>
        </mat-menu>
        <dashboard-deliveries *ngIf="can_deliver"></dashboard-deliveries>
    `,
    styles: [
        `
            :host {
                padding: 1em;
                min-width: 16rem;
            }

            [name='contact-list'] {
                padding: 0 0.5rem 0.5rem 0;
            }

            [name='contact'] {
                min-width: 8rem;
                width: 8rem;
                margin-left: 0.5rem;
                margin-top: 0.5rem;
            }

            [name='contact-name'] {
                font-weight: 400;
            }

            [name='dot'] {
                height: 8px;
                width: 8px;
                border-radius: 8px;
                background-color: #22992e;
            }

            button[name='close'] {
                position: absolute;
                display: none;
                top: 0.25rem;
                right: 0.25rem;
            }

            [name='contact']:hover button[name='close'] {
                display: initial;
            }

            button[name='contact'] {
                background-color: #f0f0f0;
                border: none;
                min-width: calc(100% - 0.5rem) !important;
            }
        `,
    ],
})
export class DashboardContactsComponent extends BaseClass {
    public readonly contacts = this._state.contacts;

    public readonly search_results = this._state.search_results;

    public readonly addUser = (u) => this._state.addContact(u);

    public readonly removeUser = (u) => this._state.removeContact(u);

    public readonly updateSearch = (s) => this._state.setOptions({ search: s });

    public search_str: string = '';

    @ViewChild('input') private _input_el: ElementRef;

    public get can_deliver() {
        return !!this._settings.get('app.can_deliver');
    }

    constructor(
        private _state: DashboardStateService,
        private _settings: SettingsService
    ) {
        super();
    }

    public ngOnInit() {
        this._state.updateContacts();
    }

    public focusInput() {
        this.timeout('focus', () => {
            if (this._input_el) {
                this._input_el.nativeElement.focus();
            }
        });
    }

    public clearInput() {
        this.search_str = '';
        this._state.setOptions({ search: '' });
    }
}
