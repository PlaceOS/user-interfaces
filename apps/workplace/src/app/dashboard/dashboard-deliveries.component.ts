import { Component } from '@angular/core';
import { MatDialog } from '@angular/material/dialog';
import { first } from 'rxjs/operators';
import { DashboardDeliveryModalComponent } from './dashboard-delivery-modal.component';

@Component({
    selector: 'dashboard-deliveries',
    template: `
        <h3 class="m-0 my-2 font-medium text-xl">Deliveries</h3>
        <div class="bg-white shadow p-2 rounded">
            <div class="divide-y" *ngIf="deliveries?.length; else empty_state">
                <div
                    class="flex items-center py-2"
                    *ngFor="let item of deliveries"
                >
                    <div class="flex-1 pl-2">
                        <div class="">
                            {{ item?.name || 'Untitled Delivery' }}
                        </div>
                        <div class="text-xs opacity-60">
                            {{
                                item && item.received
                                    ? (item.received | date: 'medium')
                                    : 'Not received'
                            }}
                        </div>
                    </div>
                    <div
                        class="m-4 h-2 w-2 rounded-full"
                        [class.bg-pending]="!item.received"
                        [class.bg-success]="item.received"
                    ></div>
                </div>
                <button
                    mat-button
                    class="w-full mt-2"
                    (click)="registerDelivery()"
                >
                    Register Delivery
                </button>
            </div>
        </div>
        <ng-template #empty_state>
            <div class="p-8 flex items-center justify-center opacity-40">
                No registered deliveries
            </div>
        </ng-template>
    `,
    styles: [``],
})
export class DashboardDeliveriesComponent {
    public deliveries = [
        { name: 'USB-C Cables', received: Date.now() },
        { name: 'Desk Fan' },
    ];

    constructor(private _dialog: MatDialog) {}

    public async registerDelivery() {
        const ref = this._dialog.open(DashboardDeliveryModalComponent);
        const details = await Promise.race([
            ref.componentInstance.events
                .pipe(first((_) => _.reason === 'done'))
                .toPromise(),
            ref.afterClosed().toPromise(),
        ]);
        if (details.reason !== 'done') return;
        this.deliveries.push(details.metadata);
        ref.close();
    }
}
