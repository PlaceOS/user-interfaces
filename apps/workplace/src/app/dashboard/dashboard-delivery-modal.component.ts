import { Component, EventEmitter, Output } from '@angular/core';
import { FormControl, FormGroup, Validators } from '@angular/forms';
import { DialogEvent } from '@placeos/common';

@Component({
    selector: 'dashboard-delivery-modal',
    template: `
        <header class="flex items-center justify-between">
            <h3>Register Delivery</h3>
            <button mat-icon-button mat-dialog-close>
                <app-icon>close</app-icon>
            </button>
        </header>
        <main class="p-2 w-[24rem]" [formGroup]="form">
            <div>
                <label>Name<span>*</span>:</label>
                <mat-form-field appearance="outline" class="w-full">
                    <input
                        matInput
                        placeholder="Name of items being delivered"
                        formControlName="name"
                    />
                    <mat-error>Name is required</mat-error>
                </mat-form-field>
            </div>
            <div>
                <label>Description:</label>
                <mat-form-field appearance="outline" class="w-full">
                    <textarea
                        matInput
                        placeholder="Description of items being delivered"
                        formControlName="description"
                    ></textarea>
                </mat-form-field>
            </div>
            <div>
                <label>Sender:</label>
                <mat-form-field appearance="outline" class="w-full">
                    <input
                        matInput
                        placeholder="Name of sender"
                        formControlName="sender"
                    />
                    <mat-error>Sender is required</mat-error>
                </mat-form-field>
            </div>
            <div>
                <label>Priority:</label>
                <mat-form-field appearance="outline" class="w-full">
                    <mat-select placeholder="Low" formControlName="priority">
                        <mat-option>Very Low</mat-option>
                        <mat-option>Low</mat-option>
                        <mat-option>Normal</mat-option>
                        <mat-option>High</mat-option>
                        <mat-option>Very Height</mat-option>
                    </mat-select>
                </mat-form-field>
            </div>
        </main>
        <footer
            class="p-2 flex items-center justify-center border-t border-gray-100"
        >
            <button mat-button class="w-32" (click)="register()">
                Register
            </button>
        </footer>
    `,
    styles: [``],
})
export class DashboardDeliveryModalComponent {
    @Output() public events = new EventEmitter<DialogEvent>();

    public readonly form = new FormGroup({
        name: new FormControl('', [Validators.required]),
        description: new FormControl(''),
        sender: new FormControl(''),
        priority: new FormControl(''),
    });

    public register() {
        this.form.markAllAsTouched();
        if (!this.form.valid) return;
        this.events.emit({ reason: 'done', metadata: this.form.value });
    }
}
