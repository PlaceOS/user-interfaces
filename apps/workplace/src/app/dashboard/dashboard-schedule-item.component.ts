import { Component, Input } from '@angular/core';
import { Booking } from '@placeos/bookings';
import { isAfter } from 'date-fns';
import { BookingLike } from '../schedule/schedule-state.service';

@Component({
    selector: '[scheduled-item]',
    template: `
        <a
            btn
            matRipple
            class="rounded-none my-1 mx-4 w-[calc(100%-2rem)] h-24 bg-base-100 hover:shadow p-0"
            [class.opacity-50]="has_ended"
            [routerLink]="['/schedule', 'view', item?.id, type]"
        >
            <div class="flex items-center border border-base-200 rounded-lg">
                <div
                    status
                    class="h-24 w-20 flex flex-col items-center justify-center text-white leading-tight"
                    [class.bg-success]="status === 'approved'"
                    [class.bg-warning]="status === 'tentative'"
                    [class.bg-error]="status === 'declined'"
                >
                    <app-icon class="text-2xl">{{ icon }}</app-icon>
                    <div class="font-normal">
                        {{
                            item?.all_day
                                ? 'All Day'
                                : (item?.date | date: 'shortTime')
                        }}
                    </div>
                    <div class="capitalize text-xs font-medium">
                        ({{ status }})
                    </div>
                </div>
                <div class="flex-1 flex flex-col text-left p-2">
                    <div class="text-xl pl-1">
                        {{ item?.title || 'Untitled Event' }}
                    </div>
                    <div
                        class="flex items-center font-normal leading-normal space-x-1"
                    >
                        <app-icon class="text-lg">today</app-icon>
                        <span class="text-sm">{{
                            item?.date | date: 'MMMM d'
                        }}</span>
                    </div>
                    <div
                        class="flex items-center font-normal leading-normal space-x-1"
                        *ngIf="item?.location || item?.description"
                    >
                        <app-icon class="text-lg">place</app-icon>
                        <span>
                            {{
                                item?.location || item?.description
                                    | slice: 0:255
                            }}
                        </span>
                    </div>
                </div>
            </div>
        </a>
    `,
    styles: [],
})
export class DashboardScheduledItemComponent {
    @Input() public item: BookingLike;

    public get type() {
        return this.item instanceof Booking ? 'booking' : 'event';
    }

    public get has_ended() {
        return (
            !this.item ||
            this.item.id === 'date' ||
            isAfter(new Date(), this.item.date + this.item.duration * 60 * 1000)
        );
    }

    public get icon() {
        if (this.status === 'declined') return 'cancel';
        if (this.item.asset_id) {
            switch ((this.item as Booking).booking_type) {
                case 'desk':
                    return 'view_quilt';
                case 'parking':
                    return 'local_parking';
            }
        }
        return 'event';
    }

    public get status() {
        if (!this.item || this.item.id === 'date') return 'declined';
        return this.item.status;
    }
}
