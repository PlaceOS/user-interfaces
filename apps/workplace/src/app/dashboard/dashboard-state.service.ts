import { Injectable } from '@angular/core';
import {
    getModule,
    PlaceMetadata,
    PlaceVariableBinding,
    showMetadata,
    updateMetadata,
} from '@placeos/ts-client';
import { BehaviorSubject, combineLatest, of } from 'rxjs';
import {
    catchError,
    debounceTime,
    filter,
    first,
    map,
    shareReplay,
    switchMap,
} from 'rxjs/operators';
import { endOfDay } from 'date-fns';

import { BaseClass, currentUser, HashMap, unique } from '@placeos/common';
import { Space } from '@placeos/spaces';
import { CalendarEvent, queryEvents } from '@placeos/events';
import { searchStaff, User } from '@placeos/users';
import { BuildingLevel, OrganisationService } from '@placeos/organisation';
import { CalendarService } from '@placeos/calendar';

export interface DashboardOptions {
    search?: string;
}

@Injectable({
    providedIn: 'root',
})
export class DashboardStateService extends BaseClass {
    private _options = new BehaviorSubject<DashboardOptions>({});
    /**  */
    private _free_spaces = new BehaviorSubject<Space[]>([]);
    /**  */
    private _upcoming_events = new BehaviorSubject<CalendarEvent[]>([]);
    /**  */
    private _contacts = new BehaviorSubject<User[]>([]);
    /**  */
    private _level_occupancy = new BehaviorSubject<BuildingLevel[]>([]);
    /**  */
    private _occupancy_binding: PlaceVariableBinding;
    /**  */
    public free_spaces = this._free_spaces.asObservable();
    /**  */
    public upcoming_events = this._upcoming_events.asObservable();
    /**  */
    public contacts = this._contacts.asObservable();
    /**  */
    public options = this._options.asObservable();

    public readonly search_results = this._options.pipe(
        debounceTime(500),
        switchMap(({ search }) => (search ? searchStaff(search) : of([]))),
        catchError((_) => []),
        shareReplay(1)
    );
    /**  */
    public level_occupancy = this._level_occupancy.asObservable();

    constructor(
        private _calendar: CalendarService,
        private _org: OrganisationService
    ) {
        super();
        this.init();
    }

    public async init() {
        await this._org.initialised.pipe(first((_) => _)).toPromise();
        this.subscription(
            'building',
            this._org.active_building
                .pipe(filter((bld) => !!bld))
                .subscribe(() => this.updateBuildingMetadata())
        );
        if (!this._org.organisation.bindings.area_management) return;
        const binding = getModule(
            this._org.organisation.bindings.area_management,
            'AreaManagement'
        ).binding('overview');
        binding.listen().subscribe((d) => this.updateOccupancy(d || {}));
        binding.bind();
    }

    public setOptions(options: Partial<DashboardOptions>) {
        this._options.next({ ...this._options.getValue(), ...options });
    }

    public pollFreeSpaces(delay: number = 10 * 1000) {
        this.updateFreeSpaces();
        this.interval('free_spaces', () => this.updateFreeSpaces(), delay);
    }

    public stopPollingFreeSpaces() {
        this.clearInterval('free_spaces');
    }

    public pollUpcomingEvents(delay: number = 10 * 1000) {
        this.updateUpcomingEvents();
        this.interval(
            'upcoming_events',
            () => this.updateUpcomingEvents(),
            delay
        );
    }

    public stopPollingUpcomingEvents() {
        this.clearInterval('upcoming_events');
    }

    public async updateContacts() {
        const metadata: PlaceMetadata = (await showMetadata(currentUser().id, {
            name: 'contacts',
        }).toPromise()) as any;
        const list = metadata.details instanceof Array ? metadata.details : [];
        this._contacts.next(list.map((i) => new User(i)));
    }

    public async addContact(user: User) {
        let users = [...this._contacts.getValue()];
        users.push(user);
        users = unique(users, 'email');
        const metadata = await updateMetadata(currentUser().id, {
            name: 'contacts',
            description: 'Contacts for the User',
            details: users,
        }).toPromise();
        const list = metadata.details instanceof Array ? metadata.details : [];
        this._contacts.next(list.map((i) => new User(i)));
    }

    public async removeContact(user: User) {
        let users = [...this._contacts.getValue()];
        users = users.filter((u) => u.email !== user.email);
        const metadata = await updateMetadata(currentUser().id, {
            name: 'contacts',
            description: 'Contacts for the User',
            details: users,
        }).toPromise();
        const list = metadata.details instanceof Array ? metadata.details : [];
        this._contacts.next(list.map((i) => new User(i)));
    }

    private async updateOccupancy(map: HashMap<{ recommendation: number }>) {
        const levels = [...this._org.levels];
        levels.sort(
            (a, b) => map[a.id]?.recommendation - map[b.id]?.recommendation
        );
        this._level_occupancy.next(levels);
    }

    private async updateFreeSpaces() {
        if (!this._org.building) return;
        const period_start = Math.floor(new Date().valueOf() / 1000);
        const period_end = Math.floor(endOfDay(new Date()).valueOf() / 1000);
        const list = await this._calendar
            .freeBusy({
                period_start,
                period_end,
                zone_ids: this._org.building.id,
            })
            .pipe(
                map((_) =>
                    _.filter(
                        (space) =>
                            !space.availability.length ||
                            space.availability.find((_) => _.status !== 'busy')
                    )
                )
            )
            .toPromise();
        list.sort((a, b) => a.capacity - b.capacity);
        this._free_spaces.next(list);
    }

    private async updateUpcomingEvents() {
        const period_start = Math.floor(new Date().valueOf() / 1000);
        const period_end = Math.floor(endOfDay(new Date()).valueOf() / 1000);
        const events = await queryEvents({
            period_start,
            period_end,
            calendars: currentUser().email,
        }).toPromise();
        this._upcoming_events.next(events);
    }

    private async updateBuildingMetadata() {
        const building = this._org.building;
        const metadata = await showMetadata(building.id, {
            name: 'bindings',
        }).toPromise();
        if (!(metadata.details as HashMap).occupancy) return;
        const details = (metadata.details as HashMap).occupancy;
        const module = getModule(details.sys, details.module, details.index);
        if (!module) return;
        if (this._occupancy_binding) {
            this._occupancy_binding.unbind();
        }
        this._occupancy_binding = module.binding('occupancy');
        this._occupancy_binding.bind();
        this.subscription(
            'occupancy_binding',
            this._occupancy_binding.listen().subscribe((value) => {
                const levels = Object.keys(value).map((key) => ({
                    id: key,
                    ...value[key],
                }));
                levels.sort(
                    (a, b) => a.recommendation_factor - b.recommendation_factor
                );
                this._level_occupancy.next(
                    levels.map((i) => this._org.levelWithID([i.id]))
                );
            })
        );
    }
}
