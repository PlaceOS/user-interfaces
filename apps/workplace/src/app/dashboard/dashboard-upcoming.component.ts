import { Component, OnDestroy, OnInit } from '@angular/core';
import { MatDialog } from '@angular/material/dialog';
import { Space } from '@placeos/spaces';

import { MapLocateModalComponent } from '../overlays/map-locate-modal.component';
import { DashboardStateService } from './dashboard-state.service';

@Component({
    selector: 'a-dashboard-upcoming',
    template: `
        <h3 class="m-0 mb-4 font-medium text-xl">Today's Events</h3>
        <div
            name="event"
            *ngFor="let event of upcoming_events | async | slice: 0:3"
            class="flex bg-white rounded-lg relative overflow-hidden mb-4"
        >
            <div name="status" class="absolute rounded-lg"></div>
            <div name="details" class="flex-1 mr-2">
                <div name="time" class="text-sm text-bold mb-2">
                    {{ event.date | date: 'shortTime' }}
                </div>
                <div name="title" class="text-sm text-bold mb-2">
                    {{ event.title }}
                </div>
                <div name="space" class="text-xs flex items-center mb-2">
                    <app-icon
                        class="mr-2"
                        [icon]="{ class: 'material-icons', content: 'room' }"
                    ></app-icon>
                    <a
                        class="text-black"
                        [class.underline]="!!event.space"
                        [matTooltip]="event.space ? 'Locate Space' : ''"
                        (click)="event.space ? locateSpace(event.space) : ''"
                    >
                        {{
                            event.space?.display_name ||
                                event.space?.name ||
                                '&lt;No Location&gt;'
                        }}
                        {{
                            event.space?.level?.display_name ||
                            event.space?.level?.name
                                ? ', ' + event.space.level.display_name ||
                                  event.space.level.name
                                : ''
                        }}
                    </a>
                </div>
                <div name="attendees" class="text-xs flex items-center mb-2">
                    <app-icon
                        class="mr-2"
                        [icon]="{ class: 'material-icons', content: 'group' }"
                    ></app-icon>
                    {{ event.guests.length }} Guest{{
                        event.guests.length === 1 ? '' : 's'
                    }}
                </div>
                <div name="guests" class="flex space-x-2 text-sm">
                    <a-user-avatar
                        *ngFor="let guest of event.guests"
                        [user]="guest"
                    ></a-user-avatar>
                </div>
            </div>
            <a
                button
                mat-button
                *ngIf="event.meeting_link"
                class="uppercase h-12"
                [href]="event.meeting_link"
                >Join Call</a
            >
        </div>
        <p
            *ngIf="!(upcoming_events | async).length"
            class="text-dark-fade text-center w-full"
        >
            No upcoming events for today
        </p>
    `,
    styles: [
        `
            :host {
                padding: 1em;
                min-width: 16rem;
            }

            [name='time'] {
                color: #00529b;
            }

            [name='event'] {
                padding: 1rem 1rem 1rem 2rem;
            }

            [name='status'] {
                top: 1rem;
                left: -0.8rem;
                height: 2.5rem;
                width: 1.6rem;
                background-color: #00529b;
            }

            button {
                height: 3em;
            }

            app-icon {
                margin-left: -0.25rem;
            }
        `,
    ],
})
export class DashboardUpcomingComponent implements OnInit, OnDestroy {
    public readonly upcoming_events = this._state.upcoming_events;

    constructor(
        private _state: DashboardStateService,
        private _dialog: MatDialog
    ) {}

    public ngOnInit() {
        this._state.pollUpcomingEvents();
    }

    public ngOnDestroy() {
        this._state.stopPollingUpcomingEvents();
    }

    public locateSpace(space: Space) {
        this._dialog.open(MapLocateModalComponent, {
            data: {
                id: space.id,
                name: space.name,
                map_id: space.map_id,
                level: space.level,
            },
        });
    }
}
