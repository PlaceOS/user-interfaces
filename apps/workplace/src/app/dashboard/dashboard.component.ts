import { Component } from '@angular/core';
import { SettingsService } from '@placeos/common';

@Component({
    selector: '[dashboard]',
    template: `
        <topbar></topbar>
        <div class="flex-1 flex sm:flex-row flex-col-reverse h-1/2 bg-gray-200">
            <nav-menu class="relative z-10"></nav-menu>
            <div
                class="relative z-0 flex flex-col flex-1 h-1/2 sm:h-auto overflow-hidden"
            >
                <main
                    class="flex flex-1 px-0 sm:px-8 flex-wrap overflow-auto h-px sm:h-auto"
                >
                    <a-dashboard-availability
                        class="flex-1 min-w-64"
                    ></a-dashboard-availability>
                    <a-dashboard-upcoming
                        class="flex-1 min-w-64"
                    ></a-dashboard-upcoming>
                    <a-dashboard-contacts
                        *ngIf="!hide_contacts"
                        class="flex-1 min-w-64"
                    ></a-dashboard-contacts>
                </main>
            </div>
        </div>
    `,
    styles: [
        `
            :host {
                position: absolute;
                top: 0;
                left: 0;
                right: 0;
                bottom: 0;
                overflow: hidden;
                display: flex;
                flex-direction: column;
            }

            main {
                min-height: 50%;
            }

            .top {
                background-color: #007ac8;
            }

            a[button].tile {
                flex: 1;
                border-radius: 0;
                background-color: #fff;
                color: rgba(0, 0, 0, 0.85);
                border: none;
                border-right: 1px solid #ccc;
            }

            a[button].tile:last-child {
                border: none;
            }
        `,
    ],
})
export class DashboardComponent {
    constructor(private _settings: SettingsService) {}

    public get hide_contacts() {
        return !!this._settings.get('app.hide_contacts');
    }
}
