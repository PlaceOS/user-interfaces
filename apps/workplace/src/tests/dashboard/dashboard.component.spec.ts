import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';
import { MockComponent } from 'ng-mocks';

import { NavMenuComponent } from '../../app/components/nav-menu.component';
import { TopbarComponent } from '../../app/components/topbar.component';
import { DashboardAvailabilityComponent } from '../../app/dashboard/dashboard-availability.component';
import { DashboardContactsComponent } from '../../app/dashboard/dashboard-contacts.component';
import { DashboardUpcomingComponent } from '../../app/dashboard/dashboard-upcoming.component';
import { DashboardComponent } from '../../app/dashboard/dashboard.component';

describe('DashboardComponent', () => {
    let spectator: Spectator<DashboardComponent>;
    const createComponent = createComponentFactory({
        component: DashboardComponent,
        declarations: [
            MockComponent(TopbarComponent),
            MockComponent(NavMenuComponent),
            MockComponent(DashboardAvailabilityComponent),
            MockComponent(DashboardUpcomingComponent),
            MockComponent(DashboardContactsComponent),
        ],
    });

    beforeEach(() => (spectator = createComponent()));

    it('should create component', () => {
        expect(spectator.component).toBeTruthy();
    });

    it('should match snapshot', () => {
        expect(spectator.element).toMatchSnapshot();
    });
});
