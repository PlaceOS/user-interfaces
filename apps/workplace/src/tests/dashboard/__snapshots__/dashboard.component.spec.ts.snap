// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <topbar />
  <div
    class="flex-1 flex sm:flex-row flex-col-reverse h-1/2 bg-gray-200"
  >
    <nav-menu
      class="relative z-10"
    />
    <div
      class="relative z-0 flex flex-col flex-1 h-1/2 sm:h-auto overflow-hidden"
    >
      <main
        class="flex flex-1 px-0 sm:px-8 flex-wrap overflow-auto h-px sm:h-auto"
      >
        <a-dashboard-availability
          class="flex-1 min-w-64"
        />
        <a-dashboard-upcoming
          class="flex-1 min-w-64"
        />
        <a-dashboard-contacts
          class="flex-1 min-w-64"
        />
        
      </main>
    </div>
  </div>
</div>
`;
